gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala

```diff
@@ -108,13 +108,28 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
   val numaBindingInfo: GlutenNumaBindingInfo = GlutenConfig.getConf.numaBindingInfo
   val substraitPlanLogLevel: String = GlutenConfig.getConf.substraitPlanLogLevel
 
-  private var planJson: String = ""
-
-  def getPlanJson: String = {
-    if (log.isDebugEnabled() && planJson.isEmpty) {
-      logWarning("Plan in JSON string is empty. This may due to the plan has not been executed.")
+  @transient
+  private var wholeStageTransformerContext: Option[WholeStageTransformContext] = None
+
+  def substraitPlan: PlanNode = {
+    if (wholeStageTransformerContext.isDefined) {
+      // TODO: remove this work around after we make `RelNode#toProtobuf` idempotent
+      //    see `SubstraitContext#getCurrentLocalFileNode`.
+      wholeStageTransformerContext.get.substraitContext.initSplitInfosIndex(0)
+      wholeStageTransformerContext.get.root
+    } else {
+      generateWholeStageTransformContext().root
     }
-    planJson
+  }
+
+  def substraitPlanJson: String = {
+    SubstraitPlanPrinterUtil.substraitPlanToJson(substraitPlan.toProtobuf)
+  }
+
+  def nativePlanString(details: Boolean = true): String = {
+    BackendsApiManager.getTransformerApiInstance.getNativePlanString(
+      substraitPlan.toProtobuf.toByteArray,
+      details)
```

Comment:
what's the difference between using `SubstraitPlanPrinterUtil.substraitPlanToJson(substraitPlan.toProtobuf)` and `BackendsApiManager.getTransformerApiInstance.getNativePlanString`?

Reply:
@zzcclp `getNativePlanString` is used to get the plan tree string from backend, e.g., velox and ch. (see pr description for an example).  `substraitPlanToJson` is used to show the substrait plan json.

Reply:
Got it. LGTM